Framework/interface/Iterator.h

```diff
@@ -17,28 +17,42 @@ namespace panda {
       friend const_type;
 
     public:
+      typedef int difference_type;
       typedef Iterator<C, is_const> self_type;
+      typedef typename C::value_type value_type;
       typedef typename std::conditional<is_const, typename C::const_reference, typename C::reference>::type reference;
       typedef typename std::conditional<is_const, typename C::const_pointer, typename C::pointer>::type pointer;
+      typedef std::random_access_iterator_tag iterator_category;
 
       Iterator() {}
       Iterator(mutable_type const& it) : unitSize_(it.unitSize_) { ptr_.obj = it.ptr_.obj; }
+
+      // implicitly assumes unitSize_ is the same
+      self_type& operator=(const_type const& rhs) { ptr_.obj = rhs.ptr_.obj; return *this; } 
+      self_type& operator=(mutable_type const& rhs) { ptr_.obj = rhs.ptr_.obj; return *this; }
+
+      bool operator==(mutable_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
+      bool operator==(const_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
+      bool operator!=(mutable_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+      bool operator!=(const_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+
       self_type& operator++() { shiftPtr_(1); return *this; }
       self_type operator++(int) { auto itr(*this); shiftPtr_(1); return itr; }
       self_type& operator--() { shiftPtr_(-1); return *this; }
       self_type operator--(int) { auto itr(*this); shiftPtr_(-1); return itr; }
+
       self_type& operator+=(int n) { shiftPtr_(n); return *this; }
       self_type& operator-=(int n) { shiftPtr_(-n); return *this; }
+
       self_type operator+(int n) const { auto itr(*this); return (itr += n); }
       self_type operator-(int n) const { auto itr(*this); return (itr -= n); }
-      bool operator==(mutable_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
-      bool operator==(const_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
-      bool operator!=(mutable_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
-      bool operator!=(const_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+
       reference operator*() const { return *ptr_.obj; }
       pointer operator->() const { return ptr_.obj; }
+
       int operator-(self_type const& rhs) const { return ptr_.obj - rhs.operator->(); }
-      self_type operator[](int n) const { return this->operator+(n); }
+      self_type& operator[](int n) { auto itr(*this); return (itr += n); }
```

Comment:
I was still unsure about this but I think I know what it should be now. In this [example from GNU libstdc++](https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a02021.html), the return type of `__normal_iterator::operator[]` is `reference`. Indeed if you think of a normal pointer as a random access iterator,
```
char* s = "hello world";
char& c = s[4]; // <- c is 'o'
```
works. And the idea is to replicate this behavior with `iterator::operator[]`.

So long story short, this line should be
```
reference operator[](int n) { auto itr(*this); return *(itr += n); }
```

Reply:
In that case, it's safe to make this function `const`, right? It's no longer possible to do assignment through this subscript (although you can modify the value directly, but I don't think that violates the `const`-ness of the iterator).

Reply:
It looks like this is what `__normal_iterator` does:
```C++
       reference
       operator[](const difference_type& __n) const
       { return _M_current[__n]; }
```

Reply:
Yes it can be (and is in the example) const indeed.